src/main/java/org/opensearch/sdk/BaseExtension.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.sdk;
+
+import java.util.Collection;
+import java.util.Collections;
+
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.xcontent.NamedXContentRegistry;
+import org.opensearch.env.Environment;
+import org.opensearch.threadpool.ThreadPool;
+
+public abstract class BaseExtension implements Extension {
+    protected SDKClient client;
+    protected ClusterService clusterService;
+    protected ThreadPool threadPool;
+    protected NamedXContentRegistry xContentRegistry;
+    protected Environment environment;
+
+    /**
+     * Empty constructor to fulfill abstract class requirements
+     */
+    public BaseExtension() {
+
+    }
+
+    @Override
+    public Collection<Object> createComponents(
+        SDKClient client,
```

Comment:
I think this should be `OpenSearchClient` to allow the user to override with their own preferred client.  The default initialization can store the result of `initializeClient()` from an internal `SDKClient` object.

Reply:
After looking more at this, SDKClient doesn't inherit from anything, and internally contains both an `OpenSearchClient` and a `RestClient` (which I think wraps the `OpenSearchClient`).  In the long term I think I'd like this to be a `RestClient`.  In the near term to get this PR completed, let's just leave this as is.  

FYI @owaiskazi19